### PR TITLE
Support openwrt

### PR DESCRIPTION
**Pull request by LLdaniel**
This is my first testing on the road to support openwrt as well on my OpenWrt router:
```shell
 # cat /etc/os-release 
NAME="OpenWrt"
VERSION="23.05.3"
ID="openwrt"
```
```shell
# opkg install hetzner_ddns_0.2.5-1_all.ipk 
Installing hetzner_ddns (0.2.5-1) to root...
Configuring hetzner_ddns.
```
```shell
 # hetzner_ddns.sh -v
hetzner_ddns 0.2.5
```
```shell
 # cat /etc/config/hetzner_ddns.conf 
interval='60'                           # Seconds between updates / TTL value
key='<redacted>'                        # Hetzner DNS API key
domain='refractedlight.eu'              # Top level domain name
records='cloud'                         # Space separated host subdomains (@ for domain itself)
```
```shell
 # service hetzner_ddns status
running
```
```shell
 # cat /var/log/hetzner_ddns.log | grep Update
[2024-11-12 22:10:24] Update IPv4 for cloud.refractedlight.eu: 127.0.0.1 => 84.135.103.184
```

Steps I want to realize for turning this draft to a final PR:
- [X] remove extra steps to copy `release/OpenWrt` files over to `mypackages` dir
- [X]  remove symlinks in `release/OpenWrt/files` in favor of copy instruction within `Makefile`
- [X] adjust `release/OpenWrt/files` `README.md` to above changes

EDIT: task list

### CODE REVIEW

File: hetzner_ddns.sh [hunk 1]
```diff
@@ -13,7 +13,24 @@ for arg in $(seq "$#"); do
             printf '%s %s\n' "$self" "$version"
             exit 0;;
         '--help'|'-h')
-            man hetzner_ddns;
+            # man usually not available on OpenWrt (too big)
+            # use short help string in that case
+            if ! command -v man > /dev/null
+            then
+                printf '%s\n\n%s\n\n' "hetzner_dns 0.2.5 (2 August 2024) - Hetzner Dynamic DNS Daemon" "Usage hetzner_dns.sh [--daemon] [CONFIG]"
```
Comment (filiparag) [inline hetzner_ddns.sh:20]: I think that using `printf` with string formatting string is superfluous here. While it loses indentation, the snipper below is somewhat shorter and faster:
```sh
if ! command -v man > /dev/null; then
    echo "${self} ${version} - Hetzner Dynamic DNS Daemon
Usage hetzner_dns.sh [--daemon] [CONFIG]
This program runs as system service /etc/init.d/hetzner_ddns on OpenWrt.
                -h, --help     Print help and exit
                -v, --version  Print version and exit
                -d, --daemon Detach from current shell and run as a deamon
Configuration file is located at /usr/local/etc/hetzner_ddns.conf
    interval='60'                           # Seconds between updates / TTL value
    key='********************************'  # Hetzner DNS API key
    domain='example.com'                    # Top level domain name
    records='homelab media vpn'             # Space separated host subdomains (@ for domain itself)
Control service with: service hetzner_ddns status|start|stop|restart after specifying the config"
else
    man hetzner_ddns;
fi
```

File: hetzner_ddns.sh [hunk 2]
```diff
@@ -13,7 +13,24 @@ for arg in $(seq "$#"); do
...
+                printf '%s\n' "This program runs as system service /etc/init.d/hetzner_ddns on OpenWrt."
```
Comment (filiparag) [inline hetzner_ddns.sh:21]: I would put every kind of configuration here like in the [man page](https://github.com/filiparag/hetzner_ddns/blob/0.2.5/hetzner_ddns.1.man#L24-L44), but you can leave that for after the PR.

File: hetzner_ddns.sh (hunk 1 above)
Comment (LLdaniel) [reply hetzner_ddns.sh]: :grimacing: I agree, thanks for that suggestion. I tried to fix that with [a218e96](https://github.com/filiparag/hetzner_ddns/pull/18/commits/a218e968d6d4326a58d6092d5e7db56aa38a07d4)

File: hetzner_ddns.sh (hunk 2 above)
Comment (LLdaniel) [reply hetzner_ddns.sh]: Sorry, I am not sure if I got you right: 
Should I describe service control with `service start|stop` etc in more detail similar to `systemctl` example in the man page? Would be glad to fix that as well, or if it is ok for you, I leave that for you to change?

Comment (filiparag) [reply hetzner_ddns.sh]: I was thinking about systemd-specifics like ["Multiple daemon instances"](https://github.com/filiparag/hetzner_ddns/blob/0.2.5/hetzner_ddns.1.man#L42-L44) section. It should either include all of them or none to keep it consistent.

You can leave it for later, as it doesn't affect the feature you have offered to implement 🙂

**Comment by LLdaniel** [comment]
Note: First I tried to leave the `Makefile` within the repository structure. But that gave me trouble when compiling the package. Probably because I named the package `hetzner_ddns` and the `feeds.conf` also expects the Makefile within the directory `hetzner_ddns`. 
Tl;dr: Maybe it is ok to copy the `release/OpenWrt` contents to a separate directory for building the package. 

Attached is my result which I also tested on my OpenWrt router. Just remove the `.txt` suffix its actually a `.ipk`. (Tricking GitHub to let me upload such a file :sweat_smile: )
[hetzner_ddns_0.2.5-1_all.ipk.txt](https://github.com/user-attachments/files/17756893/hetzner_ddns_0.2.5-1_all.ipk.txt)


How do you usually compile the packages for the releases? Maybe I can turn the `README.md` into a GitHub action within another PR?

File: release/OpenWrt/Makefile [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+PKG_NAME:=hetzner_ddns
+PKG_VERSION:=0.2.5
+PKG_RELEASE:=1
+PKG_LICENSE:=BSD-2-Clause-license
+PKG_MAINTAINER:=Filip Parag
+
+# clone sources
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://github.com/filiparag/hetzner_ddns.git
+PKG_SOURCE_VERSION:=master
```
Comment (filiparag) [inline release/OpenWrt/Makefile:12]: I think this should be set to `0.2.5` tag, as `PKG_VERSION` is not following development branch `master`. Correct me if I'm wrong.

File: release/OpenWrt/files/hetzner_ddns [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+         procd_open_instance [hetzner_ddns]
+         procd_set_param command /usr/bin/hetzner_ddns -d /etc/config/hetzner_ddns.conf
+         
+         # respawn automatically if something died, be careful if you have an alternative process supervisor
+         # if process exits sooner than respawn_threshold, it is considered crashed and after 5 retries the service is stopped
+         # if process finishes later than respawn_threshold, it is restarted unconditionally, regardless of error code
+         # notice that this is literal respawning of the process, not in a respawn-on-failure sense
+         procd_set_param respawn ${respawn_threshold:-3600} ${respawn_timeout:-5} ${respawn_retry:-5}
+
+         procd_set_param limits core="unlimited"
```
Comment (filiparag) [inline release/OpenWrt/files/hetzner_ddns:16]: If you are applying resource limits, I suggest to go all the way: one core shall be plenty.

**Review by filiparag — APPROVED** [review]
Beside some minor nitpicks, I am happy with the additions you have made.

Have you tested some edge cases on OpenWrt? Example: user deletes the config file, starts the tool manually from the command line while the daemon is running...

**Comment by filiparag** [comment]
> Probably because I named the package hetzner_ddns and the feeds.conf also expects the Makefile within the directory hetzner_ddns. But the repo itself is also named like that.

Off the top of my head, I would suggest replacing underscore with hyphen in the package name, if it will resolve the issue. Copying the `release/OpenWrt` to other directory seems like an issue users won't expect.

> Attached is my result which I also tested on my OpenWrt router. Just remove the .txt suffix its actually a .ipk. (Tricking GitHub to let me upload such a file)

Janky solution, I like it 😆

> How do you usually compile the packages for the releases? Maybe I can turn the README.md into a GitHub action within another PR?

I didn't invest in build tooling because this project gets rarely updated, so the juice is not worth the squeeze. I manually spin up Docker containers for Linux packages and VMs for others.

If you disagree with me regarding the effort/reward ratio, feel free to open an additional PR. But I would nudge you to spend your labor on more impactful work 😉

File: release/OpenWrt/files/hetzner_ddns [hunk 5]
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh /etc/rc.common
```
Comment (filiparag) [inline release/OpenWrt/files/hetzner_ddns:1]: This file should probably be moved to root of the repository, alongside other service files, like this one for [NetBSD](https://github.com/filiparag/hetzner_ddns/blob/0.2.5/hetzner_ddns.netbsd.rc).

That way it can be installed using the combined Makefile by adding a new [target](https://github.com/filiparag/hetzner_ddns/blob/0.2.5/Makefile#L13-L28):

```shell
# OpenWrt service
sudo make openwrt-rc
```

File: release/OpenWrt/Makefile (hunk 3 above)
Comment (LLdaniel) [reply release/OpenWrt/Makefile]: Yes, of course you are right. This is fixed now with [bea11d1](https://github.com/filiparag/hetzner_ddns/pull/18/commits/bea11d13da88fc315202d75d320456a2b4a1dcd4).

File: release/OpenWrt/files/hetzner_ddns (hunk 4 above)
Comment (LLdaniel) [reply release/OpenWrt/files/hetzner_ddns]: Ah, I forgot about that. It is changed to "0" with [bea11d1](https://github.com/filiparag/hetzner_ddns/pull/18/commits/bea11d13da88fc315202d75d320456a2b4a1dcd4).

This parameter is for core dumps and for debugging creating unlimited core dump sizes seemed ok. But now as you mentioned it I would go with 0 as there is not many disk space on smaller routers. 

Or would you like to omit the value completely? That would be also an option.

**Comment by LLdaniel** [comment]
Thanks for your helpful review. 
I hope it is ok if I have more changes no to read again for you... :sweat_smile: 

> Off the top of my head, I would suggest replacing underscore with hyphen in the package name, if it will resolve the issue. Copying the release/OpenWrt to other directory seems like an issue users won't expect.

Now I went with the hyphen character. I hope this should be fine and has the advantage to let the feeds point directly into the cloned `hetzner_ddns` repo.
>I didn't invest in build tooling because this project gets rarely updated, so the juice is not worth the squeeze. I manually >spin up Docker containers for Linux packages and VMs for others.
>
>If you disagree with me regarding the effort/reward ratio, feel free to open an additional PR. But I would nudge you to >spend your labor on more impactful work 😉

Ah, I see! No, I think you have the best estimate how often you release, so it is fine for me. 

> That way it can be installed using the combined Makefile by adding a new [target](https://github.com/filiparag/hetzner_ddns/blob/0.2.5/Makefile#L13-L28):

Oh, sry, forgot about that :sweat_smile: 
If you bear with me I can try to add the procd init part as well. What do you think is the best solution for using `/usr/bin/` on openwrt instead of `/usr/local/bin/`? `/usr/local/` usually does not exist, but maybe we can be the first ones to create it? Or just re-map for openwrt?

> Have you tested some edge cases on OpenWrt? Example: user deletes the config file, starts the tool manually from the command line while the daemon is running...

Good point! Let me test such edge cases please, you are right. I also will try to be available in the future if there should be upcoming issues. I already ditched my previous dyndns service in favor of the package now and can hopefully reproduce potential issues on OpenWrt.

File: release/OpenWrt/files/hetzner_ddns (hunk 5 above)
Comment (LLdaniel) [reply release/OpenWrt/files/hetzner_ddns]: Partly fixed with [bea11d1](https://github.com/filiparag/hetzner_ddns/pull/18/commits/bea11d13da88fc315202d75d320456a2b4a1dcd4). Regarding the *hetzner_ddns* `Makefile` please see below.

**Comment by filiparag** [comment]
>  What do you think is the best solution for using /usr/bin/ on openwrt instead of /usr/local/bin/? /usr/local/ usually does not exist, but maybe we can be the first ones to create it? Or just re-map for openwrt?

I intentionaly left `/usr/local/bin` installation path in the generic Makefile, as it is the most proper installation path for OS-agnostic tools built manually (not by a package manager).  
Opinionated installation paths are left for platform-specific build scripts in the `release/` directory.

> Let me test such edge cases please, you are right. I also will try to be available in the future if there should be upcoming issues. I already ditched my previous dyndns service in favor of the package now and can hopefully reproduce potential issues on OpenWrt.

Thanks for choosing this tool. I always try to eat my own dogfood, as it is the best way to find issues a normal user would encounter.

Also, sorry for a delayed response.

**Comment by LLdaniel** [comment]
Sorry for writing so much text here, but I hope it helps to keep everything as transparent as possible. So I would ask you kindly to have a look at my changes to the `Makefile` and `README.md` please. 
If you would like to add something, please let me know or maybe you can fix that afterwards if the PR is fine so far.

>I intentionaly left /usr/local/bin installation path in the generic Makefile, as it is the most proper installation path for OS-agnostic tools built manually (not by a package manager).
Opinionated installation paths are left for platform-specific build scripts in the release/ directory.

All right, I also keep that then. :+1: 
>Thanks for choosing this tool. I always try to eat my own dogfood, as it is the best way to find issues a normal user would encounter.

Haha, you are right :smile: 

>Also, sorry for a delayed response.
No worries!
I also wanted to provide a small test snippet.

[testing.txt](https://github.com/user-attachments/files/17836376/testing.txt)

Also again here my final `.ipk`, but please feel free to build it yourself, if you want. 
[hetzner-ddns_0.2.5-1_all.ipk.txt](https://github.com/user-attachments/files/17836407/hetzner-ddns_0.2.5-1_all.ipk.txt)

**Comment by filiparag** [comment]
> So I would ask you kindly to have a look at my changes to the Makefile and README.md please.
If you would like to add something, please let me know or maybe you can fix that afterwards if the PR is fine so far.

I have added additional comments for recommended changes. You can merge the PR and I will make the modifications if you don't want to.

> Also again here my final .ipk, but please feel free to build it yourself, if you want.

I will bump the version to `0.2.6` because of the help command modification and will build the package afterwards.

File: Makefile [hunk 6]
```diff
@@ -27,6 +27,10 @@ openrc: hetzner_ddns.init
 	@mkdir -p $(prefix)/etc/init.d
 	@install -m 0755 -p hetzner_ddns.init $(prefix)/etc/init.d/hetzner_ddns
 
+openwrt-procd: hetzner_ddns
+	@mkdir -p $(prefix)/etc/init.d
+	@install -m 0755 -p hetzner_ddns $(prefix)/etc/init.d/hetzner_ddns
```
[the remark this answers is not in the document]
Comment (LLdaniel) [reply Makefile]: Now addressed with [1c8f573](https://github.com/filiparag/hetzner_ddns/pull/18/commits/1c8f573b458e376a61f96f4922f359961209e13e)

File: README.md [hunk 7]
```diff
@@ -120,6 +125,9 @@ sudo service hetzner_ddns enable
 
 # OpenRC
 sudo rc-update add hetzner_ddns
+
+# OpenWrt
+service hetzner_ddns enable
```
[the remark this answers is not in the document]
Comment (LLdaniel) [reply README.md]: Now addressed with [5149594](https://github.com/filiparag/hetzner_ddns/pull/18/commits/5149594ad2c44c90f2a049e8b3c865ea922669c9)

**Comment by LLdaniel** [comment]
> I have added additional comments for recommended changes. You can merge the PR and I will make the modifications if you don't want to.

Thanks for your input and patience. I will leave the merge up to you, because I do not have the right to do so. :+1: 

> I will bump the version to 0.2.6 because of the help command modification and will build the package afterwards.

Sure, you are right this should justify a new version tag.
